Chatkit/Repositories/JoinedRoomsRepository.swift

```diff
@@ -0,0 +1,148 @@
+import Foundation
+
+/// A repository which exposes a collection of all rooms joined by the user.
+///
+/// Construct an instance of this class using `Chatkit.createJoinedRoomsRepository(...)`
+///
+/// ## What is provided
+///
+/// The repository exposes a state with a set of `rooms` which presents the rooms that the current user
+/// is a member of.
+///
+/// ## Receiving live updates
+///
+/// In order to be notified when the contents of the `rooms` or the `state` of the connection changes,
+/// implement the `JoinedRoomsRepositoryDelegate` protocol and assign
+/// the `JoinedRoomsRepository.delegate` property.
+///
+/// ## Understanding the `state` of the repository
+///
+/// The `state` property describes the state of the live update connection, either
+///   - `.initializing`: awaiting the initial set of data, or
+///   - `.connected`: updates are flowing live, or
+///   - `.degraded`: updates may be delayed due to network problems, or
+///   - `.closed`: the connection is closed, no further updates available.
+public class JoinedRoomsRepository: JoinedRoomsRepositoryProtocol {
```

Comment:
OK, this comment is potentially worthy of a face to face discussion. 

I am wondering if there's a good (indeed any) reason why we aren't just exposing a `protocol JoinedRoomsRepository` rather than a class.   

If we do this the class/implementation can remain internal and the benefits we get are:

- Gives clients themselves the possibility of testing more easily (since they can work with their own test doubles that implements the protocol)

- Would allow clients to still use our `JoinedRoomsViewModel` with a custom version of their`JoinedRoomsRepository`

- Keeps our tests clean
  - There's already a bit of a code smell in `StubJoinedRoomsRepository.report()` to work around the fact we're exposing a class
  -  Would simplify the `JoinedRoomsViewModel` tests because they can work with a `Stub`
				
- Codebase naming strategy remains consistent.  We'd have a `public protocol JoinedRoomsRepository` and an `internal class ConcreteJoinedRoomsRepository`

Reply:
We should probably discuss this with the rest of the team and make this consistent across the whole public API.

Reply:
OK, so we both discussed this at length with @mdpye and the conclusion was to...
- Not change this in this PR as it will cause some churn
- Change to use a single public protocol in a subsequent PR (as opposed to have both a public class and a public protocol because this keeps the documentation cleaner and number of entities to a minimum)
- Do some user testing on this alternative approach and then reassess later.